From ea315ab72b37b384f28fbd084fd1ecdae259708a Mon Sep 17 00:00:00 2001
From: Jeffy Chen <jeffy.chen@rock-chips.com>
Date: Wed, 15 Jun 2022 11:02:44 +0800
Subject: [PATCH 60/98] desktop-shell: Add option to hide date in panel clock

Add clock-with-date configuration to show only time without date in
panel clock. This provides a cleaner clock display for minimal panels.

Tested with:
[shell]
clock-with-date=false

Signed-off-by: Jeffy Chen <jeffy.chen@rock-chips.com>
---
 clients/desktop-shell.c | 72 ++++++++++++++++++++++++++++-------------
 1 file changed, 49 insertions(+), 23 deletions(-)

diff --git a/clients/desktop-shell.c b/clients/desktop-shell.c
index a2ec2679..4cb054db 100644
--- a/clients/desktop-shell.c
+++ b/clients/desktop-shell.c
@@ -78,6 +78,7 @@ struct desktop {
 	int want_panel;
 	enum weston_desktop_shell_panel_position panel_position;
 	enum clock_format clock_format;
+	bool clock_with_date;
 
 	struct window *grab_window;
 	struct widget *grab_widget;
@@ -111,6 +112,7 @@ struct panel {
 	int painted;
 	enum weston_desktop_shell_panel_position panel_position;
 	enum clock_format clock_format;
+	bool clock_with_date;
 	uint32_t color;
 	double scale;
 };
@@ -161,7 +163,7 @@ struct panel_clock {
 	struct widget *widget;
 	struct panel *panel;
 	struct toytimer timer;
-	char *format_string;
+	char format_string[128];
 	time_t refresh_timer;
 };
 
@@ -174,6 +176,35 @@ struct unlock_dialog {
 	struct desktop *desktop;
 };
 
+static int
+clock_get_preferred_width(enum clock_format clock_format, bool with_date)
+{
+	int width;
+
+	switch (clock_format) {
+	case CLOCK_FORMAT_MINUTES:
+		width = 160;
+		break;
+	case CLOCK_FORMAT_MINUTES_24H:
+		width = 130;
+		break;
+	case CLOCK_FORMAT_SECONDS:
+		width = 180;
+		break;
+	case CLOCK_FORMAT_SECONDS_24H:
+		width = 160;
+		break;
+	case CLOCK_FORMAT_NONE:
+	default:
+		return 0;
+	}
+
+	if (!with_date)
+		width -= 80;
+
+	return width;
+}
+
 static void
 panel_add_launchers(struct panel *panel, struct desktop *desktop);
 
@@ -471,7 +502,7 @@ panel_clock_redraw_handler(struct widget *widget, void *data)
 	cairo_text_extents(cr, string, &extents);
 	if (allocation.x > 0)
 		allocation.x +=
-			allocation.width - spacing * 1.5 - extents.width;
+			allocation.width - spacing - extents.width;
 	else
 		allocation.x +=
 			allocation.width / 2 - extents.width / 2;
@@ -522,21 +553,24 @@ panel_add_clock(struct panel *panel)
 	clock->panel = panel;
 	panel->clock = clock;
 
+	if (panel->clock_with_date)
+		strcpy(clock->format_string, "%a %b %d, ");
+
 	switch (panel->clock_format) {
 	case CLOCK_FORMAT_MINUTES:
-		clock->format_string = "%a %b %d, %I:%M %p";
+		strcat(clock->format_string, "%I:%M %p");
 		clock->refresh_timer = 60;
 		break;
 	case CLOCK_FORMAT_SECONDS:
-		clock->format_string = "%a %b %d, %I:%M:%S %p";
+		strcat(clock->format_string, "%I:%M:%S %p");
 		clock->refresh_timer = 1;
 		break;
 	case CLOCK_FORMAT_MINUTES_24H:
-		clock->format_string = "%a %b %d, %H:%M";
+		strcat(clock->format_string, "%H:%M");
 		clock->refresh_timer = 60;
 		break;
 	case CLOCK_FORMAT_SECONDS_24H:
-		clock->format_string = "%a %b %d, %H:%M:%S";
+		strcat(clock->format_string, "%H:%M:%S");
 		clock->refresh_timer = 1;
 		break;
 	case CLOCK_FORMAT_NONE:
@@ -577,10 +611,8 @@ panel_resize_handler(struct widget *widget,
 		first_pad_h = first_pad_w = 0;
 	}
 
-	if (panel->clock_format == CLOCK_FORMAT_SECONDS)
-		w = 170 * scale;
-	else /* CLOCK_FORMAT_MINUTES and 24H versions */
-		w = 150 * scale;
+	w = clock_get_preferred_width(panel->clock_format,
+				      panel->clock_with_date) * scale;
 
 	if (horizontal)
 		x = width - w;
@@ -621,19 +653,9 @@ panel_configure(void *data,
 		break;
 	case WESTON_DESKTOP_SHELL_PANEL_POSITION_LEFT:
 	case WESTON_DESKTOP_SHELL_PANEL_POSITION_RIGHT:
-		switch (desktop->clock_format) {
-		case CLOCK_FORMAT_NONE:
-			width = 32 * panel->scale;
-			break;
-		case CLOCK_FORMAT_MINUTES:
-		case CLOCK_FORMAT_MINUTES_24H:
-		case CLOCK_FORMAT_SECONDS_24H:
-			width = 150 * panel->scale;
-			break;
-		case CLOCK_FORMAT_SECONDS:
-			width = 170 * panel->scale;
-			break;
-		}
+		width = clock_get_preferred_width(desktop->clock_format,
+						  desktop->clock_with_date);
+		width = MAX(32, width) * panel->scale;
 		break;
 	}
 	window_schedule_resize(panel->window, width, height);
@@ -695,6 +717,7 @@ panel_create(struct desktop *desktop, struct output *output)
 
 	panel->panel_position = desktop->panel_position;
 	panel->clock_format = desktop->clock_format;
+	panel->clock_with_date = desktop->clock_with_date;
 	if (panel->clock_format != CLOCK_FORMAT_NONE)
 		panel_add_clock(panel);
 
@@ -1654,6 +1677,9 @@ parse_clock_format(struct desktop *desktop, struct weston_config_section *s)
 	else
 		desktop->clock_format = DEFAULT_CLOCK_FORMAT;
 	free(clock_format);
+
+	weston_config_section_get_bool(s, "clock-with-date",
+				       &desktop->clock_with_date, true);
 }
 
 int main(int argc, char *argv[])
-- 
2.39.5

